### PR TITLE
CPM-133: import user roles with permissions

### PR DESCRIPTION
**Pull request by nmarniesse**
**Description (for Contributor and Core Developer)**

Add user roles import with permission.  
Also  update the export to add the `role` column

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    public function flushStepElements(): void
+    {
+        foreach ($this->getStepElements() as $element) {
+            if ($element instanceof FlushableInterface) {
+                $element->flush();
+            }
+        }
+    }
+
+    private function updatePermissions(RoleInterface $role, array $privileges): void
```
Comment (nmarniesse) [inline src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:299]: Maybe this logic can be moved in the SecurityBundle?

File: src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        $sid = $this->aclManager->getSid($role);
+
+        foreach ($this->aclManager->getAllExtensions() as $extension) {
+            if (static::ACL_EXTENSION_KEY !== $extension->getExtensionKey()) {
+                continue;
+            }
+
+            $rootOid = $this->aclManager->getRootOid($extension->getExtensionKey());
+            foreach ($extension->getAllMaskBuilders() as $maskBuilder) {
+                $fullAccessMask = $maskBuilder->hasConst('GROUP_SYSTEM')
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:275]: We don't have const strings for this ?

File: src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+ * During a catalog install it's another step that creates the entire Role.
+ *
+ * @author    Nicolas Marniesse <nicolas.marniesse@akeneo.com>
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+final class ImportRoleTasklet extends AbstractStep implements TrackableStepInterface, LoggerAwareInterface, StoppableStepInterface
+{
+    private const ACL_EXTENSION_KEY = 'action';
+    private const MAX_ATTEMPTS_TO_CREATE_A_ROLE = 100;
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:50]: Is it still used ?

**Review by pierallard — APPROVED** [review]

File: src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php (hunk 2 above)
Comment (nmarniesse) [reply src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:272]: nope

File: src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+ *  - string role creation (e.g. "ROLE_USER")
+ *  - add/remove permissions
+ *
+ * During a catalog install it's another step that creates the entire Role.
+ *
+ * @author    Nicolas Marniesse <nicolas.marniesse@akeneo.com>
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+final class ImportRoleTasklet extends AbstractStep implements TrackableStepInterface, LoggerAwareInterface, StoppableStepInterface
```
Comment (mmetayer) [inline src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:47]: why not use a TaskletStep here? Looks like a lot of boilerplate code to me

File: src/Akeneo/UserManagement/Component/Model/Role.php [hunk 5]
```diff
@@ -67,7 +67,7 @@ public function getLabel(): ?string
      */
     public function setRole($role): void
     {
-        $this->role = (string) strtoupper($role);
+        $this->role = preg_replace('/[^a-zA-Z0-9_]/', '', strtoupper((string) $role));
```
Comment (mmetayer) [inline src/Akeneo/UserManagement/Component/Model/Role.php:70]: why ? I don't see this method used anywhere in the PR except in a test.
Also in `\Akeneo\UserManagement\Bundle\Form\Handler\AclRoleHandler` (used by the UI) there is that line of code: 
```php
$role->setRole(strtoupper(trim(preg_replace('/[^\w\-]/i', '_', $role->getLabel()))));
```
Shouldn't we use the same reg exp?

File: src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    private function buildIndexedPrivilegeNames(array $privileges): array
+    {
+        $indexedPrivilegesNames = [];
+        foreach ($privileges as $privilege) {
+            foreach ($privilege['permissions'] as $permission) {
+                if ($permission['access_level'] !== AccessLevel::NONE_LEVEL) {
+                    $name = $privilege['id'];
+                    if (false !== strpos($name, ':')) {
+                        $name = substr($name, 1 + strpos($name, ':'));
+                    }
```
Comment (mmetayer) [inline src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:300]: shouldn't we rather do that in the array converter?

File: src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php (hunk 1 above)
Comment (mmetayer) [reply src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:256]: I would keep it here personnally (better not touch the Oro namespace anymore)

**Review by mmetayer — COMMENTED** [review]
We can discuss it IRL, but I would turn the `ImportRoleTasklet` class into a real Tasklet

File: src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php (hunk 4 above)
Comment (nmarniesse) [reply src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:47]: With a regular Tasklet we don't have access to JobRepository and EventDispatcher to add warnings/errors. We can inject them but we also have to recopy some boilerplate methods that are in AbstractTasklet. Don't know what is the best way

Comment (nmarniesse) [reply src/Akeneo/UserManagement/Component/Connector/Job/ImportRoleTasklet.php:47]: And we can't stop the job manually (maybe less prioritary?)

File: src/Akeneo/UserManagement/Component/Model/Role.php (hunk 5 above)
Comment (nmarniesse) [reply src/Akeneo/UserManagement/Component/Model/Role.php]: I put back the cold code

**Pull request merged**